index.html

```diff
@@ -1151,7 +1154,98 @@ <h2>Constrainable Properties</h2>
+  unsigned long rtpTimestamp;
+};</pre>
+      <section class="notoc">
+        <h5>Members</h5>
+        <dl class="dictionary-members" data-link-for="VideoFrameMetadata" data-dfn-for="VideoFrameMetadata">
+          <dt><dfn><code>captureTime</code></dfn> of type <span class="idlMemberType">DOMHighResTimeStamp</span></dt>
+          <dd>
+            <p>The capture timestamp of the frame relative to {{Performance}}.{{Performance/timeOrigin}}. It corresponds to
+              the [=capture timestamp=] of {{MediaStreamTrack}} video frames.
+            </p>
+          </dd>
+          <dt><dfn><code>receiveTime</code></dfn> of type <span class="idlMemberType">DOMHighResTimeStamp</span></dt>
+          <dd>
+            <p>The receive time of the corresponding encoded frame relative to {{Performance}}.{{Performance/timeOrigin}}.
+              It corresponds to the [=receive timestamp=] of {{MediaStreamTrack}} video frames.</p>
+          </dd>
+          <dt><dfn><code>rtpTimestamp</code></dfn> of type <span class="idlMemberType">unsigned long</span></dt>
+          <dd>
+            <p>The RTP timestamp of the corresponding encoded frame.  It corresponds to [=RTP timestamp=] of
+            {{MediaStreamTrack}} video frames.</p>
+          </dd>
+        </dl>
+      </section>
+      <h3>Algorithms</h3>
+      The <dfn class="abstract-op">Initialize Video Frame Timestamps From Internal MediaStreamTrack Video Frame</dfn>
+      algorithm runs with |frame| as input.
+      <ol class=algorithm>
+        <li>Set {{VideoFrame/timestamp}} from [=presentation timestamp=].</li>
+        <li>Set {{VideoFrameMetadata/captureTime}} from [=capture timestamp=] if set.</li>
+        <li>Set {{VideoFrameMetadata/receiveTime}} from [=receive timestamp=] if set.</li>
+        <li>Set {{VideoFrameMetadata/rtpTimestamp}} from [=RTP timestamp=] if set.</li>
+      </ol>
+      The <dfn class="abstract-op">Copy Video Frame Timestamps To Internal MediaStreamTrack Video Frame</dfn>
+      algorithm runs with |frame| as input.
+      <ol class=algorithm>
+        <li>Set [=presentation timestamp=] from {{VideoFrame/timestamp}}.</li>
+        <li>Set [=capture timestamp=] from {{VideoFrameMetadata/captureTime}} if [=map/exist|present=].</li>
+        <li>Set [=receive timestamp=] from {{VideoFrameMetadata/receiveTime}} if [=map/exist|present=].</li>
+        <li>Set [=RTP timestamp=] from {{VideoFrameMetadata/rtpTimestamp}} if [=map/exist|present=].</li>
+      </ol>
+    </section>
+    <section>
+      <h3>Local video capture timestamps</h3>
+      <p>
+        The user agent MUST set the [= capture timestamp =] of each video frame that is sourced from
+        {{MediaDevices/getUserMedia()}} and {{MediaDevices/getDisplayMedia()}} to its best estimate of the time that
+        the frame was captured, which MUST be in the past relative to the time the frame is appearing on the track.
+        This value MUST be monotonically increasing.
```

Comment:
At TPAC, I think there was consensus on having camera tracks timestamp == capture timestamp.
It does not seem to be the case here though I would guess we keep the idea that the same clock is used.

Reply:
As per https://jsfiddle.net/4yzmwnsL/, it seems both Chrome and Safari use the same `frame.timestamp` for cloned tracks, even though the second track was cloned 2 seconds after the first one. This seems to be in contradiction with this PR.

Instead, implementations seem to be aligned with the idea that `capture timestamp` and `presentation timestamp` are both set by the source, not by the track. 

I am also not totally clear of the difference between these two timestamps for local sources.
It seems that `the time the frame is appearing on the track` would be the`presentation timestamp`, and would be slightly greater than the `capture timestamp`.
If that is the case, we should refer to both `capture timestamp` and `presentation timestamp` in that section of the PR.

Reply:
Yeah that's right, I spoke to this on TPAC. Currently, Chrome emits absolute capture timestamps from gUM/gDM capture to the webcodecs `timestamp` slot in MSTP. We believed we could do this since the nature of the timestamp isn't really well specified in webcodecs. It was later found that there are web apps that depend on a 0-based timeline and we have heuristics to support both cases in MSTG. The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.

`presentation timestamp` is 0-based and increments by frame duration. We intend to change Chrome back to 0-based here after the PR series is landed.
`capture timestamp` is absolute & unobservable on MediaStreamTracks but there are behavior requirements stated for gDM/gUM in this PR. It becomes observable first in `VideoFrameMetadata` creation at the MSTP.

> I am also not totally clear of the difference between these two timestamps for local sources.

For gUM/gDM sources, assume frame sequence indices 0, 1, 2, i ... N. `presentation timestamp[i + 1] - presentation timestamp[i] = capture timestamp[i + 1] - capture timestamp[i] = VideoFrame.duration[i]` i.e. same clock rate, different origin.

Suggestions on how to make this clearer?

Reply:
> The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.

That is fine to me.

> `presentation timestamp` is 0-based and increments by frame duration.

This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source.
For VTG, `0` would refer to the time the first video frame is enqueued.

For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`

Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).

> It was later found that there are web apps that depend on a 0-based timeline

Maybe these web apps (and the heuristics you mentioned) could tell us whether `0` should be per track, per track's sink or per track's source.

> For gUM/gDM sources, assume frame sequence indices 0, 1, 2, i ... N. `presentation timestamp[i + 1] - presentation timestamp[i] = capture timestamp[i + 1] - capture timestamp[i] = VideoFrame.duration[i]` i.e. same clock rate, different origin.

That makes sense to me.

> Suggestions on how to make this clearer?

I would add some wording stating that `presentation timestamp` and `capture timestamp` are using the same clock and have a constant offset.

Reply:
> > `presentation timestamp` is 0-based and increments by frame duration.
> 
> This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source. For VTG, `0` would refer to the time the first video frame is enqueued.
> 
> For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`
> 
> Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).
> 

I think `0` per track source is what makes most sense. WDYT?

Reply:
> > The purpose of this PR series is to clean this whole topic up and stop using 1 field for 2 things.
> 
> That is fine to me.
> 
> > `presentation timestamp` is 0-based and increments by frame duration.
> 
> This PR assumes that `0` is per track. Meaning that track and track.clone() would not have the same definition of `0`. This seems unnatural to me. I would tend to go with `0` as the time of the first frame emitted by the source. For VTG, `0` would refer to the time the first video frame is enqueued.
> 
> For this PR, that would mean changing `which is relative to the first frame appearing on the track.` to `which is relative to the first frame emitted by the track's [[Source]].`

> I think 0 per track source is what makes most sense. WDYT?

Done in [a8fc811](https://github.com/w3c/mediacapture-extensions/pull/156/commits/a8fc811bcdc51a5c3c32932ff535454efc81c9d1).


> Or it could be `0` per track's sink, meaning that MediaStreamTrackProcessor would always provide its first frame to the web app with a timestamp of `0` (provided the web app reads the frames quickly enough).
> 
> > It was later found that there are web apps that depend on a 0-based timeline
> 
> Maybe these web apps (and the heuristics you mentioned) could tell us whether `0` should be per track, per track's sink or per track's source.

The usages we found would work great if MSTP exposes 0-based from creation.


> > Suggestions on how to make this clearer?
> 
> I would add some wording stating that `presentation timestamp` and `capture timestamp` are using the same clock and have a constant offset.

Done in [a8fc811](https://github.com/w3c/mediacapture-extensions/pull/156/commits/a8fc811bcdc51a5c3c32932ff535454efc81c9d1).